compiler/damlc/daml-stdlib-src/DA/Upgrade.daml

```diff
@@ -88,6 +88,14 @@ instance (MetaEquiv c1 c2, GenConvertible f1 f2) => GenConvertible (M1 i1 c1 f1)
 instance (GenConvertible a1 a2, GenConvertible b1 b2) => GenConvertible (a1 :*: b1) (a2 :*: b2) where
   cv ~(P1 a b) = P1 (cv a) (cv b)
 
+-- product embeddings
+instance (GenConvertible a1 a2) => GenConvertible a1 (a2 :*: Opt b s) where
+  cv a = P1 (cv a) genNone
+
+-- product projections
+instance GenConvertible a1 a2 => GenConvertible (a1 :*: Opt b1 s) a2 where
```

Comment:
I'm not convinced by this instance. As it stands right now, you can remove a field from a record and then the value get's dropped during an upgrade. That doesn't sound too nice to me. Instead, it should only work if the value we drop is `None` and fail if we're trying to drop a `Some _`.

Reply:
Not sure, that would mean it's impossible to drop fields even if you wanted to. On the other side, it is, of course, dangerous if fields just get dropped unintentionally.  I'll change it for now and we'll have to see how that plays out in practice.

Reply:
actually not sure if we can do this without having a partial pattern match, because whether it's `None`, `Some _ ` is not at the type-level.

Reply:
We could make it fail at runtime if it’s a `Some`. Right now I would probably just remove the instance.

Reply:
removing the instance makes rollbacks of previously embedded data impossible, so that's not an option. I changed it to fail at runtime. But I think this is really wrong. If a user removes a field of a contract then the only thing that can possibly happen, is that the field gets deleted. We shouldn't prevent a user from doing that.

Reply:
We're not preventing people from doing that. We're just forcing them to write it down explicitly when they want to do that. I think that's fine.